sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/aio/_link_async.py

```diff
@@ -115,108 +116,116 @@ def from_incoming_frame(cls, session, handle, frame):
-                await self._outgoing_attach()
-                await self._set_state(LinkState.ATTACH_SENT)
-        elif self._session.state == SessionState.DISCARDING:
-            await self._remove_pending_deliveries()
-            await self._set_state(LinkState.DETACHED)
+        with self._lock:
+            if self._session.state == SessionState.MAPPED:
+                if not self._is_closed and self.state == LinkState.DETACHED:
+                    await self._outgoing_attach()
+                    await self._set_state(LinkState.ATTACH_SENT)
+            elif self._session.state == SessionState.DISCARDING:
+                await self._remove_pending_deliveries()
+                await self._set_state(LinkState.DETACHED)
 
     async def _outgoing_attach(self):
-        self.delivery_count = self.initial_delivery_count
-        attach_frame = AttachFrame(
-            name=self.name,
-            handle=self.handle,
-            role=self.role,
-            send_settle_mode=self.send_settle_mode,
-            rcv_settle_mode=self.rcv_settle_mode,
-            source=self.source,
-            target=self.target,
-            unsettled=self.unsettled,
-            incomplete_unsettled=self.incomplete_unsettled,
-            initial_delivery_count=self.initial_delivery_count if self.role == Role.Sender else None,
-            max_message_size=self.max_message_size,
-            offered_capabilities=self.offered_capabilities if self.state == LinkState.ATTACH_RCVD else None,
-            desired_capabilities=self.desired_capabilities if self.state == LinkState.DETACHED else None,
-            properties=self.properties
-        )
-        if self.network_trace:
-            _LOGGER.info("-> %r", attach_frame, extra=self.network_trace_params)
-        await self._session._outgoing_attach(attach_frame)
+        with self._lock:
+            self.delivery_count = self.initial_delivery_count
+            attach_frame = AttachFrame(
+                name=self.name,
+                handle=self.handle,
+                role=self.role,
+                send_settle_mode=self.send_settle_mode,
+                rcv_settle_mode=self.rcv_settle_mode,
+                source=self.source,
+                target=self.target,
+                unsettled=self.unsettled,
+                incomplete_unsettled=self.incomplete_unsettled,
+                initial_delivery_count=self.initial_delivery_count if self.role == Role.Sender else None,
+                max_message_size=self.max_message_size,
+                offered_capabilities=self.offered_capabilities if self.state == LinkState.ATTACH_RCVD else None,
+                desired_capabilities=self.desired_capabilities if self.state == LinkState.DETACHED else None,
+                properties=self.properties
+            )
+            if self.network_trace:
+                _LOGGER.info("-> %r", attach_frame, extra=self.network_trace_params)
+            await self._session._outgoing_attach(attach_frame)
```

Comment:
Would this cause an issue b/c the lock for session and the lock for connection are 2 different locks?